Enable high resolution UI artwork and remove DisableWindowsDPIScaling.

pchote opened the pull request:
This PR takes the groundwork laid by #17489, #17494, #17495, #17504, #17506, #17509, #17511 and gives us some rewarding features to justify the hard slog. The changes here should noticeably improve rendering quality on systems with more than 96 / 100% DPI. This also implements most of the remaining work towards #10382.

The first few commits fix issues that become obvious at fractional scales, and allows UI sprites (icons, color pickers) to be cleanly rendered (which should also help with #16709). The final two commits enable our new higher resolution UI artwork, and finally removes the `DisableWindowsDPIScaling` workaround that made everything blurry for many Windows players.

This can be tested on Windows by running the game with different system DPI scales. It can be tested on Linux by launching with fractional `OPENRA_DISPLAY_SCALE` values (e.g. 1.5). Testing on macOS can show that we correctly switch between different sheets when moving the window between standard and Retina Displays.

Followup PRs will enable higher resolution badge artwork (which needs to be coordinates with forum changes), add the UI Scale option to the settings, and enable automatic DPI detection on Linux.

pchote committed:
Use full window resolution on Windows.

pchote committed:
Use antialiasing filter when rendering UI icons and actors.

pchote committed:
Add antialiasing support for paletted sprites.

pchote committed:
Support rendering at non-integer display scales:

* 2x and 3x DPI artwork can be specified using
  Image2x and Image3x in chrome.yaml.
* Images are rendered using bilinear interpolation.
* For non-integer screen scales, prefer downscaling
  the next biggest resolution image over upscaling.

pchote committed:
Allocate 1px empty margin between sprites.

pchote commented on OpenRA.Game/Graphics/SpriteFont.cs:
```diff
@@ -65,48 +69,83 @@ void PrecacheColor(Color c, string name)
 			var cosa = (float)Math.Cos(-angle);
 			var sina = (float)Math.Sin(-angle);
 
+			// TODO: round positions to screen pixels as above
```
This TODO has already been resolved. I will remove it when addressing any other review comments.

pchote commented:
These commits have already had a fair amount of testing on the rendertest builds that I have been sharing on IRC and Discord [[1](https://github.com/pchote/OpenRA/releases/tag/release-20191117-renderertest1), [2](https://github.com/pchote/OpenRA/releases/tag/release-20191117-rendertest2), [3](https://github.com/pchote/OpenRA/releases/tag/release-20191117-rendertest3)], so I am confident that there are no obvious issues on any specific OS.

pchote commented:
```
21:38 <+pchote> btw i realized that i'm going to need to make a small but fundamental change to #17581 before its ready to review
21:41 <+pchote> with the legacy scaling mode the OS automatically scales the window sizes we ask for by the scaling factor, so what the game things is 1024x768 
                actually ends up giving a 1280x960 window in "real" pixels
21:41 <+pchote> which is good, because this is how things should work
21:42 <+pchote> but when we turn off the legacy scaling it turns that off, so when we ask for 1024x768 we get 1024x768, which has an effective resolution ingame of 
                only 819x614 - smaller than our minimum requirement
21:43 <+pchote> so need to change the window size code to build in the scaling factor on our side instead
```

pchote committed:
Fix font rendering at non-integer display scales.

pchote commented:
Rebased and updated. This should now be ready to review.

Mailaender commented:
Tested with custom fonts in a custom mod https://github.com/Mailaender/OpenHV and works fine. :+1:

Mailaender commented on OpenRA.Platforms.Default/Sdl2PlatformWindow.cs:
```diff
@@ -120,6 +118,25 @@ public Sdl2PlatformWindow(Size requestWindowSize, WindowMode windowMode, int bat
+				{
+					float scale = 1;
+					var scaleVariable = Environment.GetEnvironmentVariable("OPENRA_DISPLAY_SCALE");
+					if (scaleVariable != null && float.TryParse(scaleVariable, out scale))
```
This probably needs an invariant culture as testing on German locale laptop crashes unexpectedly:

```
OPENRA_DISPLAY_SCALE=1.5 ./launch-game.sh 
Platform is Linux
Engine version is 6497d1ed5d81e6be209e464e007f35d0773b1877
Using SDL 2 with OpenGL renderer
Desktop resolution: 1920x1080
No custom resolution provided, using desktop resolution
Using resolution: 20490x11520
Using window scale 15,00
Exception of type `System.InvalidOperationException`: Can not create OpenGL context. (Error: Invalid window)
  at OpenRA.Platforms.Default.Sdl2GraphicsContext.InitializeOpenGL () [0x000c1] in /home/matthias/Entwicklung/OpenHV/engine/OpenRA.Platforms.Default/Sdl2GraphicsContext.cs:53 
  at OpenRA.Platforms.Default.ThreadedGraphicsContext.RenderThread (System.Object contextObject) [0x0002a] in /home/matthias/Entwicklung/OpenHV/engine/OpenRA.Platforms.Default/ThreadedGraphicsContext.cs:79 
  at System.Threading.ThreadHelper.ThreadStart_Context (System.Object state) [0x00025] in /home/abuild/rpmbuild/BUILD/mono-6.4.0.198/mcs/class/referencesource/mscorlib/system/threading/thread.cs:78 
  at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00071] in /home/abuild/rpmbuild/BUILD/mono-6.4.0.198/mcs/class/referencesource/mscorlib/system/threading/executioncontext.cs:968 
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00000] in /home/abuild/rpmbuild/BUILD/mono-6.4.0.198/mcs/class/referencesource/mscorlib/system/threading/executioncontext.cs:910 
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state) [0x0002b] in /home/abuild/rpmbuild/BUILD/mono-6.4.0.198/mcs/class/referencesource/mscorlib/system/threading/executioncontext.cs:899 
  at System.Threading.ThreadHelper.ThreadStart (System.Object obj) [0x0000f] in /home/abuild/rpmbuild/BUILD/mono-6.4.0.198/mcs/class/referencesource/mscorlib/system/threading/thread.cs:93 

[ERROR] FATAL UNHANDLED EXCEPTION: System.InvalidOperationException: Can not create OpenGL context. (Error: Invalid window)
  at OpenRA.Platforms.Default.Sdl2GraphicsContext.InitializeOpenGL () [0x000c1] in /home/matthias/Entwicklung/OpenHV/engine/OpenRA.Platforms.Default/Sdl2GraphicsContext.cs:53 
  at OpenRA.Platforms.Default.ThreadedGraphicsContext.RenderThread (System.Object contextObject) [0x0002a] in /home/matthias/Entwicklung/OpenHV/engine/OpenRA.Platforms.Default/ThreadedGraphicsContext.cs:79 
  at System.Threading.ThreadHelper.ThreadStart_Context (System.Object state) [0x00025] in /home/abuild/rpmbuild/BUILD/mono-6.4.0.198/mcs/class/referencesource/mscorlib/system/threading/thread.cs:78 
  at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00071] in /home/abuild/rpmbuild/BUILD/mono-6.4.0.198/mcs/class/referencesource/mscorlib/system/threading/executioncontext.cs:968 
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00000] in /home/abuild/rpmbuild/BUILD/mono-6.4.0.198/mcs/class/referencesource/mscorlib/system/threading/executioncontext.cs:910 
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state) [0x0002b] in /home/abuild/rpmbuild/BUILD/mono-6.4.0.198/mcs/class/referencesource/mscorlib/system/threading/executioncontext.cs:899 
  at System.Threading.ThreadHelper.ThreadStart (System.Object obj) [0x0000f] in /home/abuild/rpmbuild/BUILD/mono-6.4.0.198/mcs/class/referencesource/mscorlib/system/threading/thread.cs:93
```

but works fine using:
```
matthias@linux-tuxedo:~/Entwicklung/OpenHV> OPENRA_DISPLAY_SCALE=1,5 ./launch-game.sh  
Platform is Linux
Engine version is 6497d1ed5d81e6be209e464e007f35d0773b1877
Using SDL 2 with OpenGL renderer
Desktop resolution: 1920x1080
No custom resolution provided, using desktop resolution
Using resolution: 2049x1152
Using window scale 1,50
OpenGL version: 3.2.0 NVIDIA 440.44
Using default sound device
Cannot connect to server socket err = Datei oder Verzeichnis nicht gefunden
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
Internal mods:
        hv: OpenHV ({DEV_VERSION})
        d2k: Dune 2000 (6497d1ed5d81e6be209e464e007f35d0773b1877)
        ts: Tiberian Sun (6497d1ed5d81e6be209e464e007f35d0773b1877)
        all: All mods (6497d1ed5d81e6be209e464e007f35d0773b1877)
        cnc: Tiberian Dawn (6497d1ed5d81e6be209e464e007f35d0773b1877)
        ra: Red Alert (6497d1ed5d81e6be209e464e007f35d0773b1877)
        modcontent: Mod Content Manager (6497d1ed5d81e6be209e464e007f35d0773b1877)
External mods:
        d2k-{DEV_VERSION}: Dune 2000 ({DEV_VERSION})
        cnc-{DEV_VERSION}: Tiberian Dawn ({DEV_VERSION})
        ts-release-20191117: Tiberian Sun (release-20191117)
        ra-df61b59746978a7187e0c5f39f7d584e249805a7: Red Alert (df61b59746978a7187e0c5f39f7d584e249805a7)
        ra-{DEV_VERSION}: Red Alert ({DEV_VERSION})
        hv-{DEV_VERSION}: OpenHV ({DEV_VERSION})
Loading mod: hv
System.Net.Sockets.SocketOptionName 0x17 is not supported at IP level
[2020-01-21T21:24:31] Game started
```

pchote replied:
This is just moving existing code, but yes, we can fix it here.

pchote replied:
Actually, it makes more sense to defer this fix to the next PR, which rewrites this code again.

pchote replied:
Fix included in 3b3a81cecc, which is one of the PRs that depends on this one.

pchote committed:
Define window size in effective pixel coordinates.

pchote commented:
Updated to fix a regression with the effective window size calculation.

pchote committed:
Extract load screen sheet handling into a superclass.

pchote committed:
Fix scroll event position calculation.

pchote committed:
Fix software cursor rendering at fractional scales.

pchote commented:
Support power icons (both player and spectator) and annotation renderables (pips, repair glyph, etc) are all missing the antialiasing scale filter. I'm a bit disappointed that nobody noticed this issue with all the test builds I have put out :cry:

pchote committed:
Use antialiasing filter when rendering world annotations.

pchote commented:
Rebased and fixed. There were some non-trivial changes to the cursor handling thanks to #17592, and the antialiasing filter is now used for annotations and the two support power palettes.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/f423d19f41d9c6220f6f7cc95658f4b094323078...a4f8daddf22b47995018b46c3d1cc2ad31001b56)
